solr/core/src/java/org/apache/solr/handler/component/SlowNodeDetector.java

```diff
@@ -0,0 +1,249 @@
+   * @param slowNodeTtl slow node list entry expire on write(detection) in millisec
+   */
+  private SlowNodeDetector(
+      double latencyDropRatioThreshold,
+      int maxSlowResponsePercentage,
+      int minShardCountPerRequest,
+      int slowLatencyThreshold,
+      long slowNodeTtl) {
+    this.latencyDropRatioThreshold = latencyDropRatioThreshold;
+    this.maxSlowResponsePercentage = maxSlowResponsePercentage;
+    this.minShardCountPerRequest = minShardCountPerRequest;
+    this.slowLatencyThreshold = slowLatencyThreshold;
+
+    Caffeine<Object, Object> builder = Caffeine.newBuilder();
+
+    if (slowNodeTtl >= 0) {
+      builder.expireAfterWrite(slowNodeTtl, java.util.concurrent.TimeUnit.MILLISECONDS);
+    }
+    slowNodes = builder.<String, Object>build().asMap();
+  }
+
+  Set<String> getSlowNodes() {
+    return new HashSet<>(slowNodes.keySet());
+  }
+
+  /** For test only */
+  void setSlowNodes(Set<String> slowNodes) {
+    this.slowNodes.clear();
+    for (String slowNode : slowNodes) {
+      this.slowNodes.put(slowNode, Boolean.TRUE);
+    }
+  }
+
+  void notifyRequestStats(RequestStats stats) {
+    Set<String> newSlowNodes = computeSlowNodes(stats);
+
+    if (newSlowNodes != null) {
+      for (String slowNode : newSlowNodes) {
+        slowNodes.put(slowNode, Boolean.TRUE);
+      }
+    }
+  }
+
+  private Set<String> computeSlowNodes(RequestStats stats) {
+    if (stats.responseLatencies.size() < minShardCountPerRequest) {
+      return null; // not enough responses to make a decision
+    }
+    int maxSlowResponseCount = stats.responseLatencies.size() * maxSlowResponsePercentage / 100;
+    if (maxSlowResponseCount < 1) {
+      return null; // not enough responses to make a decision
+    }
+
+    Collections.sort(stats.responseLatencies);
+
+    if (stats.responseLatencies.get(0).latency < slowLatencyThreshold) {
+      return null; // fastest response is not slow enough to consider any node as slow
+    }
+
+    Long previousLatency = null;
+    boolean foundLatencyDrop = false;
+    Map<String, Integer> iteratedResponseCountByNode = new HashMap<>();
+
+    int index = 0;
+    for (RequestStats.NodeLatency current : stats.responseLatencies) {
+      if (index++
+          > maxSlowResponseCount) { // too many potential slow responses, not a good data as we
+        // assume they are minority
+        break;
+      }
+      if (previousLatency != null
+          && (double) current.latency / previousLatency < latencyDropRatioThreshold) {
+        // found the drop in latencies, all the iterated nodes are potentially slow
+        foundLatencyDrop = true;
+        break;
+      }
+
+      // no latency drop point found so far and the rest latencies would not be significant enough
+      // to form a drop
+      if (current.latency < slowLatencyThreshold) {
+        break;
+      }
+
+      previousLatency = current.latency;
```

Comment:
Have you consider `previousLatency` to max latency? I feel like if we see half of the max latency then there is issue.

Reply:
If i understand correctly, are you suggesting if we find any latencies than is < 0.5 * max latencies, then flag it (flag the max?) 

Would you mind to give a solid example so I understand your proposal better? 😊 

From my investigation so far, having latencies half of the max latency is pretty common and normal. For example in this [ticket](https://fullstory.atlassian.net/browse/VAL-9270), the top latencies of various nodes are
![image](https://github.com/user-attachments/assets/2b67f123-28b7-4dce-8f77-3f24716003f0)

while the median is 15sec for all nodes. 

Therefore this PR tries to locate the "sudden drop" of latencies causes by slow node(s) by iterating the sorted numbers.

Reply:
Have many shards/node you want to consider in request? will that take effect in this case?

Reply:
The detection logic and the interrupt logic have different criteria right now

For detection, currently `DEFAULT_MIN_CORE_PER_REQUEST = 512` , which means a request has to at least have 512 cores in order trigger slow node detection. `DEFAULT_MAX_SLOW_NODE_PERCENTAGE = 10` controls how many requests should be iterated to detect slow node. This is to avoid false detection when a majority of nodes seem to be slow as other fast nodes were fast simply due to cache/no docs etc. 

For interrupt, there's no shards/nodes limit explicitly. However, there should be at least one "fast" response before the interrupt logic be triggered, so it has to be > 1  (which is true to all orgs anyway afaik? 😁 )